WIP [CA-1388] A11y: added @axe-core/react package for Accessibility linting

andrewmarcus opened the pull request:
This is a work-in-progress PR, to be taken over by the Broad team.

It installs the `axe-core/react` plugin, which, once configured properly, should be able to identify whenever a rendered page is not in compliance with [WCAG's accessibility guidelines](https://www.w3.org/WAI/standards-guidelines/wcag/). It does this by writing warnings to the console which include the severity of the violation.

Ideally, this would be hooked into the CircleCI integration tests process, to let developers know when they must make accessibility changes to their features.

This plugin does NOT identify places where enhancements semantics would benefit users but are not strictly required. These determinations must still be made by developers. For example:
* given an arbitrary set of links, aXe can't determine whether they function semantically as tabs, menus, a list of links or something else.
* given a button which causes another component to be drawn or not, aXe can't determine that it's actually a collapsible region and that the button should have an `aria-expanded` state to indicate this to users
* given an element with role `tablist`, aXe can't determine whether the widget supports navigating the tabs with the arrow keys.

## June 10 status

aXe is running on every page, but there seems to be a timing issue causing some false positives. It tells me "Document should have one main landmark" and "Page should contain a level-one heading", before the page has finished rendering. I expect this will turn out to just need some adjustment with when exactly in React's rendering cycles the plugin is called.

andrewmarcus committed:
Added @axe-core/react package and wired it up to run accessibility checks automatically outside of production mode

andrewmarcus commented:
> What if you adjust the debounce parameter to something absurd like 30000? Do the timing issues go away?

Yes, that should work I would think. It didn't seem to make a difference when I tried it, but I stopped working on this branch with the intent of handing it over.

dmohs commented:
Isaac worked on this a while ago as well. It might be wise to check the configuration parameters he used to see if they apply here: https://github.com/DataBiosphere/terra-ui/compare/iz-react-axe

petesantos commented:
@gpolumbo-broad Closing for now, feel free to open if your team picks this up.

cahrens commented:
FYI, I have updated this code on https://github.com/DataBiosphere/terra-ui/tree/a11y/car_axe-core

andrewmarcus commented:
Cool, glad to hear it!

Let me know if you need anything from me.
--
Andrew


From: Christina Ahrens Roberts ***@***.***>
Date: Monday, August 15, 2022 at 10:56 AM
To: DataBiosphere/terra-ui ***@***.***>
Cc: Andrew Marcus ***@***.***>, Author ***@***.***>
Subject: [EXT] - Re: [DataBiosphere/terra-ui] WIP [CA-1388] A11y: added @axe-core/react package for Accessibility linting (#2542)
EXTERNAL

FYI, I have updated this code on https://github.com/DataBiosphere/terra-ui/tree/a11y/car_axe-core<https://usg02.safelinks.protection.office365.us/?url=https%3A%2F%2Fgithub.com%2FDataBiosphere%2Fterra-ui%2Ftree%2Fa11y%2Fcar_axe-core&data=05%7C01%7C%7C95903bfd54f6417ca91a08da7ece6064%7C10c0370427c34ea18c2b2d9417d7ae50%7C0%7C0%7C637961722104887210%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=mDOQTvzEFxr%2Bp1Xzdw3xmD7mVEUzcWxKuaeK1KGkfnM%3D&reserved=0>

—
Reply to this email directly, view it on GitHub<https://usg02.safelinks.protection.office365.us/?url=https%3A%2F%2Fgithub.com%2FDataBiosphere%2Fterra-ui%2Fpull%2F2542%23issuecomment-1215113842&data=05%7C01%7C%7C95903bfd54f6417ca91a08da7ece6064%7C10c0370427c34ea18c2b2d9417d7ae50%7C0%7C0%7C637961722104887210%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=QCocHDuJrELbjG3zytMqn0DWHhsKGn7dW7oKxyPt9fo%3D&reserved=0>, or unsubscribe<https://usg02.safelinks.protection.office365.us/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAALTLV2SRUN766CN6I4CH7DVZJLC5ANCNFSM46PXBHWA&data=05%7C01%7C%7C95903bfd54f6417ca91a08da7ece6064%7C10c0370427c34ea18c2b2d9417d7ae50%7C0%7C0%7C637961722104887210%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=D9lbJ8cNPnfHopQDhgBkCSc6hhdWajIYDzPKTTf8%2Fo0%3D&reserved=0>.
You are receiving this because you authored the thread.Message ID: ***@***.***>
CAUTION: This email originated from outside of the organization. Do not click links, reply, or open attachments unless you recognize the sender (and sending email address) and know the content is safe.

________________________________
The information contained in this e-mail and any attachments from BlueHalo may contain confidential and/or proprietary information, and is intended only for the named recipient to whom it was originally addressed. If you are not the intended recipient, any disclosure, distribution, or copying of this e-mail or its attachments is strictly prohibited. If you have received this e-mail in error, please notify the sender immediately by return e-mail and permanently delete the e-mail and any attachments.
